framework: compute patch in a single step

kopiczko opened the pull request:
As we noticed in our operators we often do not really need update, or
want to postpone implementation. E.g. we do not support rolling
certificates, but we want to add that feature later. This require being
able to specify that update part should not be invoked by the framework.

Introduction of Patch struct is a solution to that problem. Create,
Delete and Update functions are called by the framework only when patch
has corresponding parts set.

The Patch is computed in similar way to previous GetUpdateState
function, which is now NewUpdatePatch. NewUpdatePatch function creates
a Patch object when CRO is created or updated. There is also
NewDeletePatch function which is called when CRO is deleted. Both
NewUpdatePatch and NewDeletePatch create Patch struct which in turn is
processed by Create, Delete and Update functions of the Resource.

### framework changes

ProcessAdd is redirected to ProcessUpdate.

ProcessUpdate has following flow:

    current = resource.GetCurrentState(cro)
    desired = resource.GetDesiredState(cro)
    patch = resource.NewUpdatePatch(cro, current, desired)
    patch.hasCreate && resource.Create(cro, patch.getCreate)
    patch.hasDelete && resource.Delete(cro, patch.getDelete)
    patch.hasUpdate && resource.Update(cro, patch.getUpdate)

ProcessDelete has following flow:

    current = resource.GetCurrentState(cro)
    patch = resource.NewDeletePatch(cro, current)
    patch.hasCreate && resource.Create(cro, patch.getCreate)
    patch.hasDelete && resource.Delete(cro, patch.getDelete)
    patch.hasUpdate && resource.Update(cro, patch.getUpdate)

xh3b4sd commented:
Thanks for your efforts here. I like this a lot. IMO it is a good step forward into the right direction. My only nitpick would be the method naming of the resource interface. I would like to have this. 

```
type Resource interface {
    GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error)
    GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error)

    GetUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
    GetDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)

    ProcessCreatePatch(ctx context.Context, obj, createPatch interface{}) error
    ProcessUpdatePatch(ctx context.Context, obj, createPatch interface{}) error
    ProcessDeletePatch(ctx context.Context, obj, createPatch interface{}) error
}
```

kopiczko commented:
My proposition on final naming:

```
type Resource interface {
    GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error)
    GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error)

    NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)
    NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*Patch, error)

    ApplyCreate(ctx context.Context, obj, createPatch interface{}) error
    ApplyUpdate(ctx context.Context, obj, updatePatch interface{}) error
    ApplyDelete(ctx context.Context, obj, deletePatch interface{}) error
}
```

I like ` Get*State` because state is retrieved from created resources so `get `verb fits here.
I like `New*Patch` because patch is just constructed  from current and desired state so `new` verb fits here.
I like `Apply*` because `apply` is a verb from git you use for applying the patch.

I don't want to use `Apply*Patch` because it's easy to confuse for example `ApplyUpdatePatch` with `NewUpdatePatch` which aren't directly related. I'd be fine with `Apply*PatchPart` if common suffix is important.

xh3b4sd commented:
I am also fine with `Apply*Patch`. I do not see the potential for confusion. The interface is quite different. For me it is important to have the subject which the process is about is defined. That is why only `Apply*` does not work for me that well. Lets get `Apply*Patch` in and then we can kick tires together. <3

kopiczko commented:
Open question: do we want to pass `desiredState` to `NewDeletePatch`? I don't think it's needed there, but on the other hand it's nice to keep function interfaces aligned. E.g. this may be helpful when passing them as arguments. I have mixed feelings.

xh3b4sd commented:
Lets keep the interfaces aligned and also inject the desired state into `NewDeletePath`. In most cases it should be good enough to only delete what is currently there, but when you want or have to be super defensive you could also make sure that the nuances between current and desired state are considered when cleaning up resources.

kopiczko commented:
I updated the interface according to agreed layout.
I added tests locking patch processing.
I updated resource wrappers.

CI is happy. This is ready for final review.

I assigned many folks (not only Tim who is present on every PR review, how he does that?) as reviewers on purpose. I think this is an important change affecting way we write the code. I encourage you to leave your opinions @fgimenez @JosephSalisbury @rossf7 @jgsqware @MarcelMue.

fgimenez commented on framework/framework.go:
```diff
@@ -175,10 +163,40 @@ func (f *Framework) NewCacheResourceEventHandler() *cache.ResourceEventHandlerFu
+	// thread safe. This is important because the source of truth for an
+	// operator are the reconciled resources. In case we would run the
+	// operator logic in parallel, we would run into race conditions.
+	f.mutex.Lock()
```
Nice!

fgimenez commented on framework/framework.go:
```diff
@@ -175,10 +163,40 @@ func (f *Framework) NewCacheResourceEventHandler() *cache.ResourceEventHandlerFu
+
+	f.logger.Log("action", "start", "component", "operatorkit", "function", "ProcessUpdate")
+
+	err := ProcessUpdate(ctx, obj, f.resources)
```
Nitpick, this line and the next could be merged

xh3b4sd replied:
Yeah, and I hate this. :D

fgimenez commented on framework/framework.go:
```diff
@@ -277,18 +263,46 @@ func ProcessDelete(ctx context.Context, obj interface{}, resources []Resource) e
-			return microerror.Mask(err)
+		createState, ok := patch.getCreate()
+		if ok {
+			err := r.ApplyCreatePatch(ctx, obj, createState)
```
ditto about err

fgimenez commented on framework/framework.go:
```diff
@@ -277,18 +263,46 @@ func ProcessDelete(ctx context.Context, obj interface{}, resources []Resource) e
+		}
+		deleteState, ok := patch.getDelete()
+		if ok {
+			err := r.ApplyDeletePatch(ctx, obj, deleteState)
```
ditto about err

fgimenez commented on framework/framework.go:
```diff
@@ -364,33 +380,44 @@ func ProcessUpdate(ctx context.Context, obj interface{}, resources []Resource) e
-			return microerror.Mask(err)
+		createState, ok := patch.getCreate()
+		if ok {
+			err := r.ApplyCreatePatch(ctx, obj, createState)
```
ditto about err

fgimenez commented on framework/framework.go:
```diff
@@ -364,33 +380,44 @@ func ProcessUpdate(ctx context.Context, obj interface{}, resources []Resource) e
-			return microerror.Mask(err)
+		deleteState, ok := patch.getDelete()
+		if ok {
+			err := r.ApplyDeletePatch(ctx, obj, deleteState)
```
ditto about err

fgimenez commented on framework/framework.go:
```diff
@@ -364,33 +380,44 @@ func ProcessUpdate(ctx context.Context, obj interface{}, resources []Resource) e
-			return microerror.Mask(err)
+		updateState, ok := patch.getUpdate()
+		if ok {
+			err := r.ApplyUpdatePatch(ctx, obj, updateState)
```
ditto about err

fgimenez commented on framework/spec.go:
```diff
@@ -29,81 +81,52 @@ type Resource interface {
-	// Underlying can be used for proper identification.
-	Underlying() Resource
+
+	// NewUpdatePatch is callend upon observed CRO/TPO change. It receives
```
I'm not sure if referencing CRO and TPO (or TPR later) is ok,TPOs are already deprecated and CRO may change in the future, IMO custom resource/object is more generic and difficult to get stale, WDYT?

xh3b4sd commented on framework/framework.go:
```diff
@@ -175,10 +163,40 @@ func (f *Framework) NewCacheResourceEventHandler() *cache.ResourceEventHandlerFu
 }
 
-// ProcessCreate is a drop-in for an informer's AddFunc. It receives the custom
+// ProcessAdd is a drop-in for an informer's AddFunc. It receives the custom
```
Why did this change? The godoc does not match the method name anymore.

xh3b4sd commented on framework/spec.go:
```diff
@@ -2,10 +2,62 @@ package framework
+const (
+	patchCreate = "create"
+	patchDelete = "delete"
+	patchUpdate = "update"
```
I guess this is supposed to be a `patchType`. Right now it is a `string`.

xh3b4sd commented on framework/spec.go:
```diff
@@ -2,10 +2,62 @@ package framework
+
+func (p *Patch) SetCreate(create interface{}) { p.data[patchCreate] = create }
+func (p *Patch) SetDelete(delete interface{}) { p.data[patchDelete] = delete }
+func (p *Patch) SetUpdate(update interface{}) { p.data[patchUpdate] = update }
```
I would move the patch magic to a `patch.go`. `spec.go` for me was always interfaces and structures, not logic. IIRC I did this myself a couple of times though, but it never felt right.

rossf7 commented on framework/spec.go:
```diff
@@ -2,10 +2,62 @@ package framework
+)
+
+// Patch is a set of information required in order to reconcile to the desired
+// state. Patch is split to three parts: create, delete and update. The parts
```
split into

xh3b4sd commented:
What is holding us back from merging this?

kopiczko commented:
Thanks for all the reviews. I think I addressed all the comments.

@xh3b4sd PTALA.

xh3b4sd commented on framework/patch.go:
```diff
@@ -0,0 +1,42 @@
+
+func (p *Patch) SetCreate(create interface{}) { p.data[patchCreate] = create }
+func (p *Patch) SetDelete(delete interface{}) { p.data[patchDelete] = delete }
+func (p *Patch) SetUpdate(update interface{}) { p.data[patchUpdate] = update }
```
Do we want to align the naming with the resource and call the `Set*Patch`. I would like to do this. When you are implementing a resource and you are doing `p.SetDelete` within `r.NewDeletePatch` my heart cries.

kopiczko replied:
Does your heart cry when you do `p.SetCreatePatch` and then `p.SetUpdatePatch` within `r.NewUpdatePatch`? What about calling `p.SetUpdatePatch` inside `r.NewDeletePatch`? That also can happen. Those are separate things. One is about creating `*Patch` struct and second about filling the `*Patch` struct with actual data. I don't like repetition so `Patch.SetCreatePatch` doesn't really looks like a good name for me, I'd rather call it `Patch.SetCreateData` if you want some common suffix.

Tell me what you think. I change it to w/e you want, even to `Patch.SetCreatePatch`.

xh3b4sd replied:
> Does your heart cry when you do `p.SetCreatePatch` and then `p.SetUpdatePatch` within `r.NewUpdatePatch`?

No, this would make sense to me. This was what I asked for. 

> What about calling `p.SetUpdatePatch` inside `r.NewDeletePatch`?

I wondered about this initially but it makes sense to me. Most operators will just not do it because they do not need to, but being able to apply updates on delete events makes total sense even though I cannot think about a real use case from the top of my head right now. 

> I'd rather call it `Patch.SetCreateData`

Then lets rather stick to the `State` term and lets call it `Set*State`. What about this? Calling `p.SetCreateState` within `r.NewDeletePatch` is reasonable for me.

kopiczko replied:
Then also `Apply*State`?

xh3b4sd replied:
That would probably be the consequence.

kopiczko replied:
I think I got the right name :D I think patch isn't about state, patch is about change. So maybe `Set*Change` and `Apply*Change`?

That would be a nice interface IMO:

```
Patch {
   getCreateChange() : interface{}, bool
   getDeleteChange() : interface{}, bool
   getUpdateChange() : interface{}, bool

   SetCreateChange(create interface{})
   SetDeleteChange(delete interface{})
   SetUpdateChange(update interface{})
}
```

```
Resource {
   GetCurrentState(ctx context.Context, obj interface{}) : interface{}, error
   GetDesiredState(ctx context.Context, obj interface{}) : interface{}, error

   NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) : *Patch, error
   NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) : *Patch, error

   ApplyCreateChange(ctx context.Context, obj, createPatch interface{}) : error
   ApplyDeleteChange(ctx context.Context, obj, deletePatch interface{}) : error
   ApplyUpdateChange(ctx context.Context, obj, updatePatch interface{}) : error
}
```

This nicely describes the flow: a patch is created from the diff of current and desired state. And then the changes described by the patch are applied.

xh3b4sd replied:
I do not like the naming of `Change` but I like the bigger picture. Lets go with that then.

kopiczko replied:
Done.

xh3b4sd commented:
Lets go with this.
